fastcov.py

```diff
@@ -255,13 +275,14 @@ def scanExclusionMarkers(fastcov_json, jobs, exclude_branches_sw, include_branch
 
 def distillFunction(function_raw, functions):
     function_name = function_raw["name"]
+    execution_count = int(function_raw["execution_count"])
```

Comment:
Ah, if you are converting to `int` here, it probably means gcov passes as `float` in JSON sometimes? Strange

Reply:
No, this was more likely because this (and others) are converted to / saved as strings by fastcov, which then does not sort properly.
I suggest keeping this (and others) just as ints from the beginning (there is a lot of casting to/from strings).

Reply:
The only time counts are converted to strings is when it is serializing to output with `.format()`. Counts are always stored in the format gcov passes, which as far as I can tell is `int`. I intended counts to always be ints from the beginning. 

Line numbers are stored as strings because JSON only supports strings as keys, and fastcov internally stores coverage like this:
```
{
    "sources": {
        "my_source.cpp": {
            "functions": {
                ...
            }
            "branches": {
                "69": [
                    0,
                    166
                ],
                "71": [
                    64,
                    70,
                    32
                ],
                "76": [
                    32,
                    32
                ]
            },
            "lines": {
                "37": 200,
                "40": 166,
                "42": 166,
                "46": 162,
                "48": 162,
                "49": 162,
                "51": 162,
                "61": 264
            }
        }
    }
}
```

The alternative would be to store line coverage as an 

"lines": [
    {
        "line_num": 37,
        "execution_count": 200
    },
    {
        "line_num": 40,
        "execution_count": 166
    }]

Reply:
The only time counts are converted to strings is when it is serializing to output with `.format()`. Counts are always stored in the format gcov passes, which as far as I can tell is `int` (I've been unable to reproduce counts showing up as floats). I intended counts to always be integers from the beginning. 

Line numbers are stored as strings because JSON only supports strings as keys, and fastcov internally stores coverage like this:
```
{
    "sources": {
        "my_source.cpp": {
            "functions": {
                ...
            }
            "branches": {
                "69": [
                    0,
                    166
                ],
                "71": [
                    64,
                    70,
                    32
                ],
                "76": [
                    32,
                    32
                ]
            },
            "lines": {
                "37": 200,
                "40": 166,
                "42": 166,
                "46": 162,
                "48": 162,
                "49": 162,
                "51": 162,
                "61": 264
            }
        }
    }
}
```

The alternative would be to store line coverage as an array:
```
"lines": [
    {
        "line_num": 37,
        "execution_count": 200
    },
    {
        "line_num": 40,
        "execution_count": 166
    }
]
```

However, this second way makes it much more slow/cumbersome to exclude coverage data from the final report as you now have to spin through arrays anytime you want to remove line or branch coverage instead of simply deleting a key

Reply:
Ok, I've confirmed that gcov is the one passing `float` type numbers sometimes as counts. Here is a snippet of raw JSON coming from gcov:

```
{'branches': [], 'count': 194674000.0, 'line_number': 392, 'unexecuted_block': False, 'function_name': '_ZNSt11char_traitsIcE3eofEv'}
```

So fastcov needs to explicitly cast all counts as `int` as you have done.

Reply:
@marxin - Just FYI it looks like `gcov` is sometimes passing `count` values as `float` instead of `int` in the JSON

Reply:
Although you know what? Python supports integer-as-key, so I could have fastcov store line numbers as integer keys and then they would only turn into strings upon JSON serialization. That could simplify things.

Reply:
You can go ahead and make all line number keys integers (which removes need for casting to/from int).

Or I can do it if you are done with this branch and I can merge it.

Reply:
> @marxin - Just FYI it looks like `gcov` is sometimes passing `count` values as `float` instead of `int` in the JSON

That's strange. I haven't seen `gcov` producing a float values:
```
{"gcc_version": "9.1.1 20190723 [gcc-9-branch revision 273734]", "files": [{"lines": [{"branches": [], "count": 1, "line_number": 1, "unexecuted_block": false, "function_name": "main"}, {"branches": [], "count": 1, "line_number": 3, "unexecuted_block": false, "function_name": "main"}, {"branches": [], "count": 1, "line_number": 4, "unexecuted_block": false, "function_name": "main"}], "functions": [{"blocks": 2, "end_column": 1, "start_line": 1, "name": "main", "blocks_executed": 2, "execution_count": 1, "demangled_name": "main", "start_column": 1, "end_line": 5}], "file": "main.c"}], "format_version": "1", "current_working_directory": "/tmp", "data_file": "main.c"}
```

Reply:
It only seems to happen when `count` is huge. Let me see if I can make a minimal example.

Reply:
Ok, I figured out what gcov is doing. Take the following snippet:

```
int main(int argc, char const *argv[])
{
    int x = 0;
    for (int i = 0; i < 12345678; i++) {
        x += i/2;
    }
    return 0;
}
```
It looks like anything over say, a million, get put into scientific json notation:

`{"branches": [], "count": 1.23457e+07, "line_number": 8, "unexecuted_block": false, "function_name": "main"}`

Which is interpreted as floating point by python:

`{'branches': [], 'count': 12345700.0, 'line_number': 8, 'unexecuted_block': False, 'function_name': 'main'}`

It may just be the particular JSON library gcov uses so you might not have control over the types it chooses during serialization. fastcov going forward will explicitly cast as `int`. 

Also note slight loss in accuracy for large counts (`12345678` vs `12345700`) - not that it's a big deal since I doubt anyone needs that kind of accuracy for code coverage line counts.

Reply:
> It looks like anything over say, a million, gets put into scientific json notation:

Haha, you are right. I'll take a look tomorrow.

Reply:
There's a patch candidate:
https://gcc.gnu.org/ml/gcc-patches/2019-07/msg01854.html